tests/PHPStan/Analyser/nsrt/bug-12902.php

```diff
@@ -0,0 +1,72 @@
+<?php declare(strict_types = 1); // lint >= 8.1
+
+namespace Bug12902;
+
+use function PHPStan\Testing\assertNativeType;
+use function PHPStan\Testing\assertType;
+
+class NarrowsNativeReadonlyUnion {
+	private readonly int|float $i;
+
+	public function __construct()
+	{
+		$this->i = getInt();
+		assertType('int', $this->i);
+		assertNativeType('int', $this->i);
+	}
+
+	public function doFoo(): void {
+		assertType('int', $this->i);
+		assertNativeType('int', $this->i);
+	}
+}
+
+class NarrowsNativeUnion {
+	private int|float $i;
+
+	public function __construct()
+	{
+		$this->i = getInt();
+		assertType('int', $this->i);
+		assertNativeType('int', $this->i);
+
+		$this->impureCall();;
+		assertType('float|int', $this->i);
+		assertNativeType('float|int', $this->i);
+	}
+
+	public function doFoo(): void {
+		assertType('float|int', $this->i);
+		assertNativeType('float|int', $this->i);
+	}
+
+	/** @phpstan-impure  */
+	public function impureCall(): void {}
+}
+
+class NarrowsStaticNativeUnion {
+	private static int|float $i;
+
+	public function __construct()
+	{
+		self::$i = getInt();
+		assertType('float|int', self::$i); // could be int
+		assertNativeType('float|int', self::$i); // could be int
+
+		$this->impureCall();
+		assertType('float|int', self::$i);
+		assertNativeType('float|int', self::$i);
+	}
```

Comment:
the impureCall does not forget about already narrowed types of static property fetches, therefore we don't narrow in this scenario with this PR.

-> we only improve the instance-call case, because non-static fetches will be forgotten as the test in `NarrowsNativeUnion` proves